### PR TITLE
Webpack v4.0.0-alpha support

### PR DESCRIPTION
**Pull request by Graham42**
This is ~a rough go at~ the code to make tests pass for supporting webpack@4.0.0-alpha-4.

~I'm not sure that this can be merged with out major version bump since webpack now requires node version >=6 and html-webpack-plugin currently still supports  4 and 5.~
Edit: This is actually probably fine, if people are using webpack the intersection of supported node versions is >=6, I don't see a need to change the required node version for this plugin.

Status as of `webpack@4.0.0-alpha-4`:
- [x] Caching Tests passing in development mode
- [x] Example fixture Tests passing in production mode
- [x] Basics Tests passing in development mode

### CODE REVIEW

File: lib/compiler.js [hunk 1]
```diff
@@ -78,10 +78,20 @@ module.exports.compileTemplate = function compileTemplate (template, context, ou
...
+        // In webpack 4 the plugin interface changed, so check for available fns
+        var outputName = compilation.mainTemplate.applyPluginsWaterfall
+          ? compilation.mainTemplate.applyPluginsWaterfall(
+              'asset-path',
+              outputOptions.filename,
+            {
+              hash: childCompilation.hash,
+              chunk: entries[0]
+            })
+          : compilation.mainTemplate.hooks.assetPath.call(outputOptions.filename, {
```
Comment (sokra) [inline lib/compiler.js]: Use `compilation.mainTemplate.getAssetPath` instead

File: index.js [hunk 2]
```diff
@@ -71,7 +85,12 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
     // Sort chunks
     chunks = self.sortChunks(chunks, self.options.chunksSortMode);
     // Let plugins alter the chunks and the chunk sorting
-    chunks = compilation.applyPluginsWaterfall('html-webpack-plugin-alter-chunks', chunks, { plugin: self });
+    if (compilation.applyPluginsWaterfall) {
```
Comment (sokra) [inline index.js]: reverse the order here: Try new API first, fallback to old API

File: index.js [hunk 3]
```diff
@@ -313,7 +332,12 @@ HtmlWebpackPlugin.prototype.addFileToAssets = function (filename, compilation) {
   })
   .then(function (results) {
     var basename = path.basename(filename);
-    compilation.fileDependencies.push(filename);
+    if (compilation.fileDependencies.push) {
```
Comment (sokra) [inline index.js]: reverse the order here: Try new API first, fallback to old API

File: index.js [hunk 4]
```diff
@@ -640,16 +664,49 @@ HtmlWebpackPlugin.prototype.getAssetFiles = function (assets) {
...
-  return function (eventName, requiresResult, pluginArgs) {
-    return promisedApplyPluginsAsyncWaterfall(eventName, pluginArgs)
-      .then(function (result) {
-        if (requiresResult && !result) {
-          compilation.warnings.push(new Error('Using ' + eventName + ' without returning a result is deprecated.'));
-        }
-        return _.extend(pluginArgs, result);
-      });
-  };
+  if (compilation.applyPluginsAsyncWaterfall) {
```
Comment (sokra) [inline index.js]: reverse the order here: Try new API first, fallback to old API

**Comment by Graham42** [comment]
going to see if there's a better way to view the diffs for fixture tests. the current output is not easy to tell what's wrong.  :confused:

**Comment by Graham42** [comment]
Rebased to clean up the commits. Included the changes suggested from @sokra, thanks for the review. 

Tests all pass now :tada: ,  However I've made some assumptions which I'd like feedback on. 

I think the caching tests only make sense for `mode = 'development'` because that's how people will be running dev servers that require recompiling.

I think `mode = 'production'` makes sense for the fixture tests from looking a the other fixture files. I tried generating both production and development variations, and the production ones look more like what's there for webpack 3.

The basic tests I think should be run for both modes but I'm not sure.

**Comment by donaldpipowitch** [comment]
Thank you for the PR. [I just give it a try](https://github.com/donaldpipowitch/experiments/commit/d522c9bd64ec893c1daebc9a721052ef67b8f72c) and for my small example it worked without errors ❤️

**Comment by donaldpipowitch** [comment]
Thank you for the PR. I just tried it against the webpack v4 alpha in a [small example](https://github.com/donaldpipowitch/experiments/tree/master/babel-compiles-typescript) and it worked without problems. ✌️

**Comment by Graham42** [comment]
This good to go. The deprecation warnings can be fixed in a separate PR. As it is this PR as it is enables compatibility with webpack-4.0.0-alpha.4. cc @jantimon @TheLarkInn

**Comment by Graham42** [comment]
Travis is timing out on network requests, going to push again to trigger build

**Comment by alexandernst** [comment]
So... merge?

**Comment by TheLarkInn** [comment]
@Graham42 this looks great to me as well.

@jantimon would you like to release this under '@next' tag similar to how we are shipping alphas? Let me know where we can help.

**Comment by niieani** [comment]
Are there any ETAs on this getting released as an alpha/next/beta?

**Comment by Graham42** [comment]
Note that this plugin breaks with `webpack@4.0.0-alpha-5` because of reliance on old common chunk behaviour. 

Not sure when I'll have a chance to look at that, but if someone else wants to work on it, could probably build on top of this branch.

**Comment by craftgear** [comment]
I use this branch git://github.com/Graham42/html-webpack-plugin.git#feat/webpack-4 with webpack@4.0.0-beta0 and have no problem.

Why is this PR not merged yet?

**Comment by gaterking** [comment]
@craftgear 
`\node_modules\html-webpack-plugin\index.js:50
      compilation.hooks.htmlWebpackPluginAlterChunks = new SyncWaterfallHook(['chunks', 'objectWithPluginRef']);
TypeError: SyncWaterfallHook is not a constructor
\node_modules\html-webpack-plugin\index.js:50:56`

**Comment by ankibalyan** [comment]
waiting for this! <3

**Comment by andriijas** [comment]
@gaterking that happens if you get an old version of webpack installed, html-webpack-plugin uses an older version of webpack for testing and its a mess upgrading it right now. Basically extract-text-webpack-plugin#707  need to land first.

What you can do is make sure you have correct version of webpack and tapable in your node_modules otherwise npm i tapable , make sure you get the correct version

**Comment by deadcoder0904** [comment]
After hours of searching I found the issue 😂 

@jantimon Can you merge this one please❓

**Comment by TheLarkInn** [comment]
Update: currently jantimon is in Asia for work right now until first week of March. We've forked html-webpack-plugin for now to have the PR sent against there just to ensure people can actually have a chance to test and use this package prior to v4

Would someone like to take the PR and revase against webpack-contrib/html-webpack-plugin ?

**Comment by probablyup** [comment]
Some of the tests also are checking for commonschunk, which no longer exists.

**Comment by Graham42** [comment]
Continuing this PR over at https://github.com/webpack-contrib/html-webpack-plugin/pull/3

**Comment by hanxue** [comment]
After installing the changes from 

```
npm i -D git://github.com/Graham42/html-webpack-plugin.git#feat/webpack-4
```

I still get the following error

```
TypeError: SyncWaterfallHook is not a constructor
    at /a/b/c/node_modules/html-webpack-plugin/index.js:50:56
    at SyncHook.eval [as call] (eval at create (/a/b/c/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:17:12), <anonymous>:11:1)
```

Manage to resolve this error by installing the latest [tapable v1.0.0](https://github.com/webpack/tapable/releases/tag/v1.0.0)

**Comment by niemyjski** [comment]
Any updates on this?

**Comment by caseyWebb** [comment]
@niemyjski 

> 🛑 If you use HtmlWebpackPlugin
>
> For this release, we gave the ecosystem a month to upgrade any plugins or loaders to use the new webpack 4 API’s. However, Jan Nicklas has been away with work obligations and therefore, we have provided a patched fork of html-webpack-plugin . For now you can install it by doing the following:
>
> `$ yarn add html-webpack-plugin@webpack-contrib/html-webpack-plugin`
>
> When Jan returns from overseas work at the end of the month, we plan to merge our fork upstream into jantimon/html-webpack-plugin ! Until then, if you have any issues, you can submit them here!

via: https://medium.com/webpack/webpack-4-released-today-6cdb994702d4

**Comment by deadcoder0904** [comment]
@niemyjski This PR works & you can install it like that on the Medium article or [like this](https://stackoverflow.com/questions/33181297/npm-install-from-github-pull-request)

**Comment by jantimon** [comment]
Merged - ty so much @Graham42

**Comment by ankibalyan** [comment]
realeasing something? `html-webpack-plugin@4.0.0` or `html-webpack-plugin@next`?

**Comment by deadcoder0904** [comment]
Just do `npm i --save-dev html-webpack-plugin` or `yarn add --dev html-webpack-plugin`

Its been released on [npm](https://www.npmjs.com/package/html-webpack-plugin)

**Pull request closed without merging**